yt_dlp/extractor/youtube.py

```diff
@@ -4552,19 +4552,43 @@ def process_language(container, base_url, lang_code, sub_name, query):
             'uploader_id': channel_handle,
             'uploader_url': format_field(channel_handle, None, 'https://www.youtube.com/%s', default=None),
         })
+
+        def get_pacific_tz():
+            # Python 3.8 should be deprecated soon
+            # This fallback may make the timestamp slightly inaccurate for 3.8 users.
+            if sys.version_info < (3, 9):
+                return dt.timedelta(hours=-7)
+            from zoneinfo import ZoneInfo
+            return dt.datetime.now(ZoneInfo('US/Pacific')).utcoffset()
+
+        # We only want timestamp IF it has second precision
+        # Additionally, if there is no timezone present, we should assume it is in PT.
+        timestamp = (
+            parse_iso8601(get_first(microformats, 'uploadDate'), timezone=get_pacific_tz())
+            or parse_iso8601(search_meta('uploadDate'), timezone=get_pacific_tz())
+        )
+        upload_date = (
+            dt.datetime.fromtimestamp(timestamp, dt.timezone.utc).strftime('%Y%m%d') if timestamp else
+            (
+                unified_strdate(get_first(microformats, 'uploadDate'))
+                or unified_strdate(search_meta('uploadDate'))
+            ))
+
+        # In the case we cannot get the timestamp:
         # The upload date for scheduled, live and past live streams / premieres in microformats
         # may be different from the stream date. Although not in UTC, we will prefer it in this case.
         # See: https://github.com/yt-dlp/yt-dlp/pull/2223#issuecomment-1008485139
-        upload_date = (
-            unified_strdate(get_first(microformats, 'uploadDate'))
-            or unified_strdate(search_meta('uploadDate')))
         if not upload_date or (
-            live_status in ('not_live', None)
+            not timestamp
+            and live_status in ('not_live', None)
             and 'no-youtube-prefer-utc-upload-date' not in self.get_param('compat_opts', [])
```

Comment:
How should this compat opt work?

Reply:
I believe youtube-dl is also planning to implement `timestamp` as UTC. @dirkf Can you confirm? In that case, we should just make the compat option a no-op

Reply:
Mhm. It also gets messy with the fact we will be changing the scheduled, live and past live streams / premieres upload_date to be utc now, since we can accurately get it.

Reply:
Yes, my WIP extractor gets the timestamp when possible, currently believed to be always, but I need to go back and check the age-gate API data.

Per https://github.com/yt-dlp/yt-dlp/issues/9829#issuecomment-2088407474, the meta fields in the video page as well as the hydration JSON are using ISO 8601 with TZ offset. Presumably the API fields will mirror the hydration JSON.

The [example code that I posted](https://github.com/yt-dlp/yt-dlp/issues/9829#issuecomment-2090207026) included the idea of `own_upload_date` to represent the cases where the automatic core `upload_date` shouldn't be used (because it should be in Pacific time, not UTC) and the unresolved `TODO: ...` to handle the calculation of `upload_date` in Pacific time from a known `timestamp` relative to UTC.

The problem is to know which actual TZ offset from (PST, PDT) applies in Mountain View, CA, for a particular UTC time. Even in Py >= 3.9 with `zone_info`, the library docs say that there's no cross-platform solution to this sort of TZ processing. If it should be required to do that the code would need to include the DST rules for the Pacific TZ, for each year where the calculation has to be done.

I think that this really argues for "retiring" the compat option and hoping that all the data sources used by the extractor are now providing ISO 8601 data with TZ.